packages/core/contracts/proxy-scripts/lsp-broker/LspUniswapV2Broker.sol

```diff
@@ -15,27 +18,197 @@ import "../../financial-templates/long-short-pair/LongShortPair.sol";
+        TransferHelper.safeApprove(address(longToken), address(router), longToken.balanceOf(address(this)));
+        TransferHelper.safeApprove(address(shortToken), address(router), shortToken.balanceOf(address(this)));
+        router.addLiquidity(
+            address(longToken),
+            address(shortToken),
+            longToken.balanceOf(address(this)),
+            shortToken.balanceOf(address(this)),
+            0,
+            0,
+            address(this),
+            deadline
+        );
+        {
+            // 4) Finally, if the caller is an EOA, send back the LP tokens from minting + any dust that was left over.
+            // The dust is in the long and short tokens and will be left over as a result of small rounding errors in
+            // the calculation on how many long/short should be bought/sold to meet the pool ratio before minting.
+            if (callingAsEOA) {
+                // Send the LP tokens back to the minter.
+                IERC20 LPToken = IERC20(address(pairFor(router.factory(), address(longToken), address(shortToken))));
+                TransferHelper.safeTransfer(address(LPToken), msg.sender, LPToken.balanceOf(address(this)));
+
+                // Send any dust left over back to the minter.
+                TransferHelper.safeTransfer(address(longToken), msg.sender, longToken.balanceOf(address(this)));
+                TransferHelper.safeTransfer(address(shortToken), msg.sender, shortToken.balanceOf(address(this)));
+            }
+        }
+    }
+
+    // For a given mint size (m) and pool balances (ra & rb), compute the trade size such that the resulting ratio
+    // of tokens, considering the mint and trade,is equal to the pool ratio after the trade. This is computed using the
+    // following logic: Assume m tokens are minted in equal proportion of long and short. This contract will therefore
+    // hold t_l=m+△l long tokens and t_s=m-△s after a trade of s for l tokens. The ratio t_l/t_s must equal the pool
+    // ratio after the trade. The pool ratio, considering the trade can be expressed as (R_a-△a)(R_b+△bλ)=k for a
+    // trade of △a for △b and a swap fee of λ=(1-swapFee). Using this, with a bit of algebra and simplification, we can
+    // solve for the resultant ratio to be (m+△a)/(m-△b)=(R_a-△a)/(R_b+△b). i.e the ratio of mint+trade must equal the
+    // ratio of token A and B in the pool after the trade has concluded. We also know that △a can be solved for as
+    // △a=(△bλR_a)/(R_b+△bλ) by manipulating a known uniswap equation. Using this equation and the expression of ratios
+    // we can solve simaltaniously for △b. I.e how many short tokens do we need to sell such that the ratio of long and
+    // short after the trade equal the pool ratios. Numerically, this works out to this form:
+    // △b=(sqrt(R_b(R_a+m)(R_a(R_b(λ^2+2λ+1)+4λm)+R_b(λ^2-2λ+1)m))+R_b(-λ-1)(R_a+m))/(2λ(R_a+m)). For how this was solved
```

Comment:
Maybe I'm missing the subtlety here, but after reading your comment, could the same thing be computed and mostly just reuse the logic we've written elsewhere to trade to a particular price? This way could be just as complicated, so even if this is correct, I'm not necessarily saying we should change anything. Just seemed a little simpler in my head.

Note that the following is just pseudocode that is under-representative of the complexity.
```
// We know that all the minted tokens are going to end up in the
// uniswap pool regardless, so the final reserves _must_ be the following.
long_reserves_final = long_minted + long_reserves_initial;
short_reserves_final = short_minted + short_reserves_final;

// Since we know that these reserves are going to be where we end up,
// we can divide them to get the price.
// Because adding liquidity doesn't change the price,
// we know this is the price we will mint at.
desired_price = long_reserves_final / short_reserves_final;

tradeToPrice(desired_price); // What the trader DSProxy code does.
router.mint(balanceOf(short_token), balanceOf(long_token)); // Mint using all of our tokens.
```

Reply:
What you put down is pretty much the same reasoning I used to get to this form. The form I presented is numerically the same logic but has been manipulated to simplify the equations so they will fit in solidity. I'm pretty sure you will arrive at an equally complex form if you do it this way (if not more complex). We should chat about it more IRL.

Reply:
I tried implementing your technique @mrice32. it worked out to be _much_ easier to reason about and had a lot simpler math. _However_ the dust leftover is measurably worse. On a simple setup with a pool with 1000000/100000 long/short (10:1) ratio has less than 100 wei dust in long with my computation for a mint of 1000 units of collateral. Your method has about 1.3e18 in long dust (this is still less than 0.3% of the total, which would be the fee amount).

I spent a bunch of time trying to re-work this to introduce additional fees into the computation to get it to track correctly with your technique but could never get the dust anywhere close to my computation.

I think this is due to the fees being ignored at some level in this computation. However, I’m not convinced by this as the fees will end up in the respective pools, irrespective of the swap size so that’s a bit confusing.

In any case, for now, I think we should stick with my implementation as it works and has low dust, despite its high complexity. You agree?

Reply:
Agreed!